src/core/frames/frame_controller.js

```diff
@@ -299,7 +301,14 @@ export class FrameController {
 
     if (newFrameElement) {
       const snapshot = new Snapshot(newFrameElement)
-      const renderer = new FrameRenderer(this, this.view.snapshot, snapshot, FrameRenderer.renderElement, false, false)
+      let renderer
+
+      if (this.element.shouldReloadWithMorph) {
+        renderer = new MorphingFrameRenderer(this, this.view.snapshot, snapshot, MorphingFrameRenderer.renderElement, false, false)
```

Comment:
Actually, here you are changing the renderer for all cases. We only need to modify the renderer when calling the `reload` method, otherwise, any Turbo frame that has the `refresh="morph"` attribute would be rendered with morphing always, not only with page refreshes or when calling the `reload` method. Probably the simplest approach here is to reuse logic that already exists, moving it to `FrameController`.

```js
  sourceURLReloaded() {
    if (this.element.shouldReloadWithMorph) {
      this.element.addEventListener("turbo:before-frame-render", ({ detail }) => {
        detail.render = MorphingFrameRenderer.renderElement
      }, { once: true })
    }

    const { src } = this.element
    this.element.removeAttribute("complete")
    this.element.src = null
    this.element.src = src
    return this.element.loaded
  }
```

Reply:
Thanks for the feedback @brunoprietog I was hoping that would be a feature rather than a bug. :) You don't think we should support morphing for more cases? I know that for the overall page we're being really conservative right now and only morphing when a redirect happens back to the same URL. I was hoping that if we explicitly set reload="morphing" on a turbo-frame then we could morph in all cases even when the turbo-frame transitions in a way other than `.reload()`. In all my testing, the morphing logic is good enough to transition even complex page transitions. But maybe there are some limitations of the idiomorph which we're trying to stay within?

Reply:
Yes, I also think it's a good idea to support morphing in Turbo frames in more cases, but here it's more a matter of naming and expectation. `refresh="morph"`, means that when you refresh the frame you expect it to be morphed. It doesn't mean that when navigating that frame it will render with morphing necessarily. We can't assume that it works well because we've done tests. We've seen that with morphing all sorts of strange situations can occur, so I'd be a little cautious with this change.

Reply:
Ah, that makes sense.

I have incorporated your feedback in this PR. It now only morphs when an explicit .reload() is called on the page (or when the page is morphing). I think I did this all correctly. I've updated my test cases and ensured I didn't inadvertently break some other tests. In the process I discovered a mistaken test which I've fixed (and left a comment on).